File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;
using namespace std::chrono_literals;

/**
 * A node that creates a subscriber to the 'number' topic and publishes a new number
 * to the 'number_count' topic.
 */
class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter")
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10, 
            std::bind(&NumberCounterNode::callbackNumbers, this, _1));
        

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>(
            "number_count", 10);

        server_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter",
            std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
    
        RCLCPP_INFO(this->get_logger(), "Number counter has been started.");

    }

private:
    /**
     * A callback function that receives a number from the 'number' topic, 
     * adds to it the value of the counter multiplied by 2,
     * and publishes the result to the 'number_count' topic. 
     */
    void callbackNumbers(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        auto msgNumber = example_interfaces::msg::Int64();
        newNumber_ = msg->data;
        msg->data = newNumber_ + (counter_ * 2);
        counter_++;
        msgNumber.data = msg->data;
        publisher_->publish(msgNumber);
    }

    /**
     * A callback function that resets the counter to 0 if the request data is true,
     * and sends a response message indicating whether the counter was reset or not.
     */
    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                              const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data)
        {
            counter_ = 0;
            response->message = std::string("Counter has been reset.");
            response->success = true;
        }
        else
        {
            response->message = "Counter has not been reset.";
            response->success = false;
        }
        RCLCPP_INFO(this->get_logger(), "%s", response->message.c_str());
    }

    // Member variables
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
    int newNumber_;
    int counter_ = 1;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

"""
A ROS2 Node that subscribes to the 'number' topic and 
counts the number of messages received and publishes the count.
"""
class NumberCounterNode(Node):
    
    """
    A function that initializes the number counter, 
    subscription to the 'number' topic,
    and publisher for the 'number_count' topic.
    """
    def __init__(self):
        super().__init__("number_counter")
        self.subscription_ = self.create_subscription(Int64, "number", self.number_callback, 10)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.server_ = self.create_service(SetBool, "reset_counter", self.reset_counter_callback)
        self.count_ = 0
        self.get_logger().info("Number counter has been started.")
    
    """
    A callback function that increments the count each time 
    a message is received on the 'number' topic
    and logs the current count.

    param msg: Int64 - The message received from the 'number' topic.
    """
    def number_callback(self, msg):
        message = msg.data
        self.count_ += 1
        msg.data = msg.data + (self.count_ * 2)
        self.publisher_.publish(msg)

    """
    A callback function that resets the counter when 
    a request is received on the 'reset_counter' service.
    """
    def reset_counter_callback(self, request, response):
        if request.data:
            self.count_ = 0
            response.success = True
            response.message = "Counter reset successfully."
        else:
            response.success = False
            response.message = "Counter reset failed. Request data must be True."
        return response
    
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()